[WIP] Add Python Script for Coupling Global Models to Regional Models

danieldouglas92 opened the pull request:
From a discussion with @gassmoeller, @jdannberg, @alarshi  and @naliboff. May also be of interest to @tjhei!

Adds a python script which demonstrates a workflow for extracting the velocity from a global model (S2ORTS cookbook in this case) along the 4 lateral boundaries of a regional 3D spherical chunk model. The east/west boundaries are great circles, and I calculate the normal vector for the plane that defines the east/west boundaries and use the `slice` filter to extract the velocity from the global models. The north/south boundaries are not necessarily great circles, and so I use the `threshold` filter to extract cells on a constant latitude to extract the velocity from the global models. Then, I write each boundary to an ASCII file which is formatted such that it can be directly applied to an ASPECT model. The velocity from the S2ORTS cookbook along each of the 4 lateral boundaries for a regional model spanning from 20 <= latitude <= 50 and 190 <= longitude <= 230 is shown below.

![S20RTS_test](https://github.com/user-attachments/assets/c6599c0c-d69e-48a6-8767-8393501cffd2)

I have never used pvpython before this, so there may be a simpler way to do this than what I've done here, or there may even be a better package to use. Open to feedback! Also, is it worth including in this PR an example .prm file which uses the ASCII boundary files?

naliboff reviewed (dismissed):
@danieldouglas92 -Thanks for the contribution! Here are the initial comments before you push the updates we discussed.

danieldouglas92 commented:
Thanks for the feedback @naliboff! I addressed the comments and I removed the ASCII files for now. I've confirmed that the workflow works and produces ASCII files which can be used in a regional chunk model, and that the correct velocity is applied to the regional model boundaries. I show an example below.

This is showing the Paraview output for the North boundary of a regional model with the setup that I described in the opening PR message.
![image](https://github.com/user-attachments/assets/57430967-1c8c-4280-bf0c-ad3db728cd0e)

And the same boundary isolated in the global S20RTS model.
![image](https://github.com/user-attachments/assets/29560f2f-5a68-49de-8566-00f97a10ae48)

@naliboff and I discussed that the python script should be merged first, and then in a follow up PR I'll add a cookbook that goes into more detail actually taking the output from this script and applying it to an ASPECT model in a more thorough way.

gassmoeller reviewed (commented):
Looks pretty good, thanks for updating and creating the PR! Just some smaller comments.

danieldouglas92 commented:
Thanks for the review @gassmoeller!

gassmoeller commented:
@danieldouglas92 Is this ready for another look or do you want to make more changes?

danieldouglas92 commented:
@gassmoeller Nope I addressed your changes! Whenever you find time another review would be appreciated :)

naliboff commented on contrib/python/scripts/extract_local_velocity.py:
```diff
@@ -0,0 +1,314 @@
+    reader = simple.OpenDataFile(temp_datafile)
+    reader.UpdatePipeline()
+
+    # Create a TableToPoints filter
```
```suggestion
    # Create a TableToPoints filter that will be used later for storing interpolated data from the slices
```

naliboff commented on contrib/python/scripts/extract_local_velocity.py:
```diff
@@ -0,0 +1,314 @@
+    tabletopoints.Input = reader
+    tabletopoints.UpdatePipeline()
+
+    # Take the cut_slice object and interpolate the fields onto the TableToPoints filter above.
```
Are there arguments within `ResampleWithDataset` that allow specifying the interpolation method?

naliboff reviewed (commented):
@danieldouglas92 - Sorry, I forgot to submit a number of comments and a prior review. Some may now be outdated, and if that is the case please ignore them.

danieldouglas92 commented:
@naliboff thanks for getting around to the review! I addressed your comments and hopefully the inline code comments are more clear now. @gassmoeller hopefully you didn't start reviewing within the last few hours

gassmoeller commented on contrib/python/scripts/extract_local_velocity.py:
```diff
@@ -0,0 +1,333 @@
+# 5. The lateral resolution of the regional model
+# 6. The maximum and minimum radius, latitude, and  longitude for the regional model.
+
+# This shows an example for applying this script to the global model
```
See here: https://github.com/dealii/dealii/pull/17717#discussion_r1767822607. Leading this's without explanation are confusing. Where is this example?

Ah, I found it below. What you want to say is something like:
The default settings inside the script are set for an example case that applies this script ...

danieldouglas92 replied:
Not leading with `This` is great advice that makes a lot of sense once someone points it out but is something I likely never would have realized otherwise! I'll definitely have to remember this going forward

gassmoeller reviewed (commented):
Thanks, almost ready. There is one place in the documentation that is a bit confusing, please improve. Then this PR should be good to go.

gassmoeller reviewed (approved):
:+1:

gassmoeller commented:
@naliboff have your comments be addressed? Then this is good to go.

gassmoeller commented:
Let's merge this. It is useful in any case and we can still improve it later if necessary.
